sphinx/builders/linkcheck.py

```diff
@@ -318,6 +318,7 @@ def check_uri() -> tuple[str, str, int]:
                     # Read the whole document and see if #anchor exists
                     response = requests.get(req_url, stream=True, config=self.config,
                                             auth=auth_info, **kwargs)
+                    response.close()  # no HTTP body reads required; close the response
```

Comment:
Reading the content _is_ required here, as part of `check_anchor`.

(so a context-manager approach might be more appropriate here)

Reply:
A better explanation, because I forgot what was going on here:

In the general case, we want to ensure that `response.raise_for_status` and `response.close` are both called after receiving a response from the server.  They provide error-checking and resource cleanup, respectively.

Most of the time, we can call `close` as soon as the HTTP status line (200 OK, etc) is received.

The anchor-checking case (looking for a specific HTML ID within a documentation webpage) is different though: it doesn't seem to make sense to anchor-check on error responses (HTTP 4xx), even though theoretically the server _could_ include HTML and the anchors we're looking for in the response.  Documentation references shouldn't be served from pages that produce 4xx responses.

So in that case we want to call `response.raise_for_status` first.

It _might_ be possible to move the `response.close` call into the `finally` clause of the `try...except` handler.

The context manager approach provided by the `requests` library calls `response.close` when the `response` context-manager exits, though: that's convenient and could be easier than writing a `finally` clause that has to handle multiple different arrival paths.